### PR TITLE
Add camera smoothing and cinematic mode (F8)

### PR DESCRIPTION
**Pull request by rubenwardy**
Implements part of #2323

```
# Enable cinematic mode
#cinematic = false
# Camera smoothing - smooths rotation of camera. 0 is no smoothing.
# Must be equal to or greater than 0, and less than 1.
#camera_smoothing = 0.0
# Camera smoothing when in cinematic mode
#cinematic_camera_smoothing = 0.7
```

### CODE REVIEW

**Comment by rubenwardy** [comment]
[Youtube Video](http://youtu.be/wCkFRsCM0UY)

**Comment by RHRhino** [comment]
That video looks great! :+1:

**Comment by Calinou** [comment]
:+1:

**Comment by nerzhul** [comment]
I approve

**Comment by rubenwardy** [comment]
This will need checking to make sure that some fluke of C++ doesnt make this reduce the FPS, but it should be fine. It grabs a float from the setting every tick, and does `1 - float`

File: src/game.cpp [hunk 1]
```diff
@@ -1788,7 +1789,12 @@ void Game::run()
 		updateProfilers(runData, stats, draw_times, dtime);
 		processUserInput(&flags, &runData, dtime);
 		// Update camera before player movement to avoid camera lag of one frame
-		updateCameraDirection(&cam_view, &flags);
+		updateCameraDirection(&cam_view_target, &flags);
+		float cam_smoothing = 1 - g_settings->getFloat("camera_smoothing");
```
Comment (Zeno-) [inline src/game.cpp]: Either this needs to be `MYMAX(0.1f, 1 - g_settings->getFloat("camera_smoothing"); // or something similar`, or the acceptable range for cam(era)_smoothing limited in some other way.

**Comment by kilbith** [comment]
What about a dedicated key for enable/disable the _cinematic mode_ ? (eg. F8)

Youtubers are not that accustomed with the commands/.conf

**Comment by rubenwardy** [comment]
I was considering that. What if players want smoothing in normal mode, but
to a lesser degree?

If core devs agree, I could add an cinematic_mode = false setting and use
cinematic_camera_smoothing when it is true, and camera_smoothing when its
not

Please note that I will also be doing a pull request soon for smooth
movement, and that would be part of cinematic mode.

**Comment by Calinou** [comment]
Ideally, this should not be FPS-dependent.

**Comment by rubenwardy** [comment]
To do:
* Squash
* Add settings to minetest.conf.example

Tested with FPS at 2, 5, 10 and 40, and there is no real difference. It is much smoother at higher FPSs, but that's because of the jittering.

File: src/game.cpp [hunk 2]
```diff
@@ -1788,7 +1792,20 @@ void Game::run()
...
-		updateCameraDirection(&cam_view, &flags);
+		updateCameraDirection(&cam_view_target, &flags);
+		float cam_smoothing = 0;
+		if (g_settings->getBool("cinematic"))
+			cam_smoothing = 1 - g_settings->getFloat("cinematic_camera_smoothing");
+		else
+			cam_smoothing = 1 - g_settings->getFloat("camera_smoothing");
+		if (cam_smoothing <= 0)
+			cam_smoothing = 0.01;
+		if (cam_smoothing > 1)
```
Comment (Zeno-) [inline src/game.cpp]: else if (not just if... both conditions can't be true in this case)

You could also use `cam_smoothing = rangelim(cam_smoothing, 0.01f, 1.0f);` (see util/numeric.h)

Comment (nerzhul) [inline src/game.cpp]: Please use numeric.h as @Zeno- said.

**Comment by SmallJoker** [comment]
+1 That's a great feature!

**Comment by falsechicken** [comment]
:+1:  Nice!

**Comment by nerzhul** [comment]
@rubenwardy we need to have a rework on this PR before integration

**Comment by rubenwardy** [comment]
Problems with making it FPS independant:

* Exceeding 0 < cam_smoothing <= 1.
* Larger dtime also means larger mouse displacement.

It would be good if the setting has an inverse effect, so 1 - 1 / setting, this makes the numbers more representative. 1000 instead of 0.999.

**Comment by nerzhul** [comment]
For me this PR is good, tested and approved. @rubenwardy is this finished?

**Comment by rubenwardy** [comment]
It is. I can't think of a way to make it fully fps independent, but I don't think that matters anyway - larger dtime also means larger mouse displacement.

**Comment by nerzhul** [comment]
I don't see any FPS problem on intel GPU and AMD GPU

**Comment by nerzhul** [comment]
Commited, thanks

d65a90a86bf5f1422a64ff190abefb8f7f199460

**Comment by rubenwardy** [comment]
Awesome. I'd like this to be merged, then :)

**Pull request closed without merging**